architecture/replication-layer.md

```diff
@@ -0,0 +1,114 @@
+
+To achieve, this each lease renewal or transfer also attempts to collocate them. In practice, that means that the mismatch is rare and self-corrects quickly.
+
+#### Epoch-Based Leases (Table Data)
+
+To manage leases for table data, CockroachDB implements a notion of "epochs," which are defined as the period between a node joining a cluster and a node disconnecting from a cluster. When the node disconnects, the epoch is considered changed, and the node immediately loses all of it leases.
+
+This mechanism lets us avoid tracking leases for every range, which eliminates a substantial amount of traffic we would otherwise incur. Instead, we assume leases don't expire until a node loses connection.
+
+#### Expiration-Based Leases (Meta & System Ranges)
+
+Your table's meta and system ranges (detailed in the Distribution Layer) are treated as normal key-value data, and therefore have Leases, as well. However, instead of using epochs, they have an expiration-based lease. These leases simply expire at a particular timestamp (typically a few seconds)––however, as long as the node continues proposing Raft commands, it continues to extend the expiration of the lease. If it doesn't, the next node containing a replica of the range that tries to read from or write to the range will become the Lease Holder.
+
+### Membership Changes: Rebalance/Repair
+
+Whenever there are changes to a cluster's number of nodes, the members of Raft groups change and, to ensure optimal survivability and performance, replicas need to be rebalanced. What that looks like varies depending on whether the membership change is nodes being added or going offline.
+
+**Nodes added**: The new node communicates information about itself to other nodes, indicating that it has space available. The cluster then rebalances some replicas onto the new node.
+
+**Nodes going offline**: If a member of a Raft group ceases to respond, after 5 minutes, the cluster begins to rebalance by replicating the data the downed node held onto other nodes.
+
+#### Rebalancing Replicas
+
+When CockroachDB detects a membership change, ultimately, replicas are moved between nodes.
+
+This is achieved by taking using a snapshot of a replica from the Lease Holder, and then sending the data to another node over [gRPC](distribution-layer.html#grpc). After the transfer has been completed, the node with the new replica joins that range's Raft group; it then detects that its latest timestamp is behind the most recent entries in the Raft log and it replays all of the actions in the Raft log on itself.
+
+## Interactions with Other Layers
+
+### Replication & Distribution Layers
+
+The Replication Layer receives requests from its and other nodes' `DistSender`, which it accepts if it is the Lease Holder or errors if it's not. These KV requests are then turned into Raft commands.
```

Comment:
`errors if it's not` - Is that overly general? Doesn't it know who the lease is? If so, does the node forward the request to the right place, or send some type of guidance back to `DistSender`?